Establishing the connection to webinspectord can sometimes fail

pvollan opened the pull request:
#### 961481a24076e5ec9061c05aaee06770c3b00161
<pre>
Establishing the connection to webinspectord can sometimes fail
<a href="https://bugs.webkit.org/show_bug.cgi?id=276512">https://bugs.webkit.org/show_bug.cgi?id=276512</a>
<a href="https://rdar.apple.com/130712941">rdar://130712941</a>

Reviewed by Chris Dumez.

There is currently a race condition when establishing the XPC connection from the WebContent process to
webinspectord. If the XPC connection is being established after we block access to launchd, establishing
the connection will not succeed. Launchd is being blocked before first page load. This patch addresses
this by establishing the XPC connection before this point in time.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::RemoteInspector::connectToWebInspector):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/280912@main">https://commits.webkit.org/280912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb6a90cba3b22dda4a97d05038c278bfd17f5bda

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8416 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8604 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46979 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5993 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34979 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50109 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27809 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31743 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7404 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7420 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51063 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53689 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7672 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63283 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57213 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7739 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54203 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50120 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54339 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1612 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78974 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8652 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33128 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13102 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34214 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35298 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33959 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -346,6 +346,11 @@ static bool canAccessWebInspectorMachPort()
 
+void RemoteInspector::notifyCanConnectToWebInspector()
+{
+    notify_post(WIRServiceAvailableNotification);
```
what if Web Inspector is not enabled on the device?  i think this would cause us to attempt to create an XPC connection to `com.apple.webinspector` every time we finish initializing a WebProcess

should we instead `WIRServiceAvailabilityCheckNotification` to ask `com.apple.webinspector` to wake up, which would then send back a `WIRServiceAvailableNotification` if enabled?

pvollan replied:
Ah, that's a good catch! I have addressed this in the latest patch by checking whether we got a Web Inspector sandbox extension from the UI process. We only receive this sandbox extension if Web Inspector is enabled. What do you think?

Thanks for reviewing!

dcrousso commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -614,6 +614,7 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
         RetainPtr<CFDataRef> auditData = adoptCF(CFDataCreate(nullptr, (const UInt8*)&*auditToken, sizeof(*auditToken)));
         Inspector::RemoteInspector::singleton().setParentProcessInformation(WebCore::presentingApplicationPID(), auditData);
     }
+    Inspector::RemoteInspector::singleton().notifyCanConnectToWebInspector();
```
im pretty sure that Web Inspector will actually attempt to connect to `com.apple.webinspector` in the UIProcess as part of support for PSON, so is this actually necessary for Web Inspector to function?

pvollan replied:
At least in the Service Worker case, I believe we need the connection to `com.apple.webinspector`.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -346,6 +346,13 @@ static bool canAccessWebInspectorMachPort()
+void RemoteInspector::connectToWebInspector()
+{
+    dispatch_async(m_xpcQueue, ^{
+        RemoteInspector::singleton().setupXPCConnectionIfNeeded();
```
i dont think it's the case, but it might be possible that `WIRServiceAvailabilityCheckNotification` is actually necessary in order for `com.apple.webinspector` to properly be created (or woken up if it already exists)

dcrousso commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -614,6 +614,12 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
+    // This is because we block launchd before the first page load, and launchd is required to establish the connection.
+    // This is only done if Web Inspector is enabled, which is determined by the size of the Web Inspector extension vector.
+    // See WebProcessProxy::shouldEnableRemoteInspector().
+    if (parameters.enableRemoteWebInspectorExtensionHandles.size())
```
is it possible for `parameters.enableRemoteWebInspectorExtensionHandles` to be consumed before this is reached?  i see some code in `WebProcess:: platformInitializeWebProcess` that seems like it will affect this, but not entirely sure if `.size()` would still be valid even after that

pvollan replied:
That's a good point! The sandbox extensions are consumed before this point, but the vector is left unchanged, so `size()` is still valid.

pvollan committed:
Establishing the connection to webinspectord can sometimes fail
https://bugs.webkit.org/show_bug.cgi?id=276512
rdar://130712941

Reviewed by Chris Dumez.

There is currently a race condition when establishing the XPC connection from the WebContent process to
webinspectord. If the XPC connection is being established after we block access to launchd, establishing
the connection will not succeed. Launchd is being blocked before first page load. This patch addresses
this by establishing the XPC connection before this point in time.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::RemoteInspector::connectToWebInspector):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):

Canonical link: https://commits.webkit.org/280912@main

JosephPecoraro commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -346,6 +346,13 @@ static bool canAccessWebInspectorMachPort()
 
+void RemoteInspector::connectToWebInspector()
+{
+    dispatch_async(m_xpcQueue, ^{
```
Is this being "async" going to continue the race, or is it somehow guaranteed to happen before the further lockdown point?

pvollan replied:
That's a very good point! I initially thought that `setupXPCConnectionIfNeeded()` had to be called on `m_xpcQueue`, but looking closer, that does not seem to be the case. If `setupXPCConnectionIfNeeded()` can be called on the main thread, we should call it synchronously here.
